File: app/src/main/java/com/zionhuang/music/ui/widgets/MediaWidgetsController.java
package com.zionhuang.music.ui.widgets;

import android.animation.ValueAnimator;
import android.content.Context;
import android.provider.Settings;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.view.animation.LinearInterpolator;
import android.widget.ProgressBar;
import android.widget.SeekBar;
import android.widget.TextView;

import static com.zionhuang.music.utils.Utils.makeTimeString;

public class MediaWidgetsController {
    private static final String TAG = "MediaWidgetsController";
    private ProgressBar mProgressBar;
    private SeekBar mSeekBar;
    private boolean mSeekBarIsTracking;
    private TextView mProgressTextView;

    private MediaControllerCompat mMediaController;
    private ControllerCallback mControllerCallback;
    private ValueAnimator mProgressAnimator;
    private long mDuration;
    private float mDurationScale;

    public MediaWidgetsController(Context context, ProgressBar progressBar, SeekBar seekBar, TextView progressTextView) {
        mProgressBar = progressBar;
        mSeekBar = seekBar;
        mProgressTextView = progressTextView;
        mDurationScale = Settings.Global.getFloat(context.getContentResolver(), Settings.Global.ANIMATOR_DURATION_SCALE, 1f);
        mSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                mProgressTextView.setText(makeTimeString(progress / 1000));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
                mSeekBarIsTracking = true;
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                if (mMediaController != null) {
                    mMediaController.getTransportControls().seekTo(mSeekBar.getProgress());
                }
                mSeekBarIsTracking = false;
            }
        });
    }

    public void setMediaController(MediaControllerCompat mediaController) {
        if (mediaController != null) {
            mControllerCallback = new ControllerCallback();
            mediaController.registerCallback(mControllerCallback);
            mControllerCallback.onMetadataChanged(mediaController.getMetadata());
            mControllerCallback.onPlaybackStateChanged(mediaController.getPlaybackState());
        } else if (mMediaController != null) {
            mMediaController.unregisterCallback(mControllerCallback);
            mControllerCallback = null;
        }
        mMediaController = mediaController;
    }

    public void disconnectController() {
        if (mMediaController != null) {
            mMediaController.unregisterCallback(mControllerCallback);
            mControllerCallback = null;
            mMediaController = null;
        }
    }

    private class ControllerCallback extends MediaControllerCompat.Callback implements ValueAnimator.AnimatorUpdateListener {
        @Override
        public void onPlaybackStateChanged(PlaybackStateCompat state) {
            super.onPlaybackStateChanged(state);
            if (state == null) return;
            if (mProgressAnimator != null) {
                mProgressAnimator.cancel();
                mProgressAnimator = null;
            }

            int progress = (int) state.getPosition();
            mProgressBar.setProgress(progress);
            mSeekBar.setProgress(progress);
            mProgressTextView.setText(makeTimeString((long) (progress) / 1000));

            if (state.getState() == PlaybackStateCompat.STATE_PLAYING) {
                int timeToEnd = (int) ((mDuration - progress) / state.getPlaybackSpeed());
                if (timeToEnd > 0) {
                    if (mProgressAnimator != null) mProgressAnimator.cancel();
                    mProgressAnimator = ValueAnimator.ofInt(progress, (int) mDuration).setDuration((long) (timeToEnd / mDurationScale));
                    mProgressAnimator.setInterpolator(new LinearInterpolator());
                    mProgressAnimator.addUpdateListener(this);
                    mProgressAnimator.start();
                }
            }
        }

        @Override
        public void onMetadataChanged(MediaMetadataCompat metadata) {
            super.onMetadataChanged(metadata);
            if (metadata != null) {
                mDuration = metadata.getLong(MediaMetadataCompat.METADATA_KEY_DURATION);
                mProgressBar.setMax((int) mDuration);
                mSeekBar.setMax((int) mDuration);
            }
            if (mMediaController != null) {
                onPlaybackStateChanged(mMediaController.getPlaybackState());
            }
        }

        @Override
        public void onAnimationUpdate(ValueAnimator animation) {
            if (mSeekBarIsTracking) {
                animation.cancel();
                return;
            }
            int animatedValue = (int) animation.getAnimatedValue();
            mProgressBar.setProgress(animatedValue);
            mSeekBar.setProgress(animatedValue);
            mProgressTextView.setText(makeTimeString(animatedValue / 1000));
        }
    }
}

File: app/src/main/java/com/zionhuang/music/ui/widgets/MediaWidgetsController.kt
package com.zionhuang.music.ui.widgets

import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.content.Context
import android.provider.Settings
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.view.animation.LinearInterpolator
import android.widget.ProgressBar
import android.widget.TextView
import com.google.android.material.slider.Slider
import com.zionhuang.music.utils.makeTimeString

class MediaWidgetsController(
    context: Context,
    private val progressBar: ProgressBar,
    private val slider: Slider,
    private val progressTextView: TextView,
) {
    private var sliderIsTracking = false
    private var mediaController: MediaControllerCompat? = null

    private var controllerCallback: ControllerCallback? = null
    private var progressAnimator: ValueAnimator? = null
    private var duration: Long = 0
    private val durationScale: Float = Settings.Global.getFloat(context.contentResolver, Settings.Global.ANIMATOR_DURATION_SCALE, 1f)

    init {
        slider.addOnSliderTouchListener(object : Slider.OnSliderTouchListener {
            override fun onStartTrackingTouch(slider: Slider) {
                sliderIsTracking = true
            }

            override fun onStopTrackingTouch(slider: Slider) {
                mediaController?.transportControls?.seekTo(slider.value.toLong())
                sliderIsTracking = false
            }

        })
        slider.addOnChangeListener { _, value, _ ->
            progressTextView.text = makeTimeString((value).toLong())
        }
    }

    fun setMediaController(newController: MediaControllerCompat?) {
        if (newController != null) {
            controllerCallback = ControllerCallback().also {
                newController.registerCallback(it)
                it.onMetadataChanged(newController.metadata)
                it.onPlaybackStateChanged(newController.playbackState)
            }
        } else if (mediaController != null) {
            if (controllerCallback != null) {
                mediaController!!.unregisterCallback(controllerCallback!!)
                controllerCallback = null
            }
        }
        mediaController = newController
    }

    fun disconnectController() {
        if (mediaController != null) {
            mediaController!!.unregisterCallback(controllerCallback!!)
            controllerCallback = null
            mediaController = null
        }
    }

    private inner class ControllerCallback : MediaControllerCompat.Callback(), AnimatorUpdateListener {
        override fun onPlaybackStateChanged(state: PlaybackStateCompat?) {
            super.onPlaybackStateChanged(state)
            state ?: return

            progressAnimator?.cancel()
            progressAnimator = null

            val progress = state.position.toInt()
            progressBar.progress = progress
            if (slider.isEnabled) {
                slider.value = progress.toFloat().coerceIn(slider.valueFrom, slider.valueTo)
            }
            progressTextView.text = makeTimeString(progress.toLong())
            if (state.state == PlaybackStateCompat.STATE_PLAYING) {
                val timeToEnd = ((duration - progress) / state.playbackSpeed).toInt()
                if (timeToEnd > 0) {
                    progressAnimator?.cancel()
                    progressAnimator = ValueAnimator.ofInt(progress, duration.toInt()).apply {
                        duration = (timeToEnd / durationScale).toLong()
                        interpolator = LinearInterpolator()
                        addUpdateListener(this@ControllerCallback)
                        start()
                    }
                }
            }
        }

        override fun onMetadataChanged(metadata: MediaMetadataCompat) {
            super.onMetadataChanged(metadata)
            duration = metadata.getLong(MediaMetadataCompat.METADATA_KEY_DURATION)
            progressBar.max = duration.toInt()
            if (duration <= 0L) {
                slider.isEnabled = false
            } else {
                slider.isEnabled = true
                slider.valueTo = duration.toFloat()
            }
            mediaController?.let {
                onPlaybackStateChanged(it.playbackState)
            }
        }

        override fun onAnimationUpdate(animation: ValueAnimator) {
            if (sliderIsTracking) {
                animation.cancel()
                return
            }
            val animatedValue = animation.animatedValue as Int
            progressBar.progress = animatedValue
            slider.value = animatedValue.toFloat().coerceIn(slider.valueFrom, slider.valueTo)
            progressTextView.text = makeTimeString(animatedValue.toLong())
        }
    }
}